kedro/templates/project/{{ cookiecutter.repo_name }}/pyproject.toml

```diff
@@ -1,3 +1,53 @@
+[build-system]
+requires = ["setuptools"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "{{ cookiecutter.python_package }}"
+dependencies = [
+    "kedro~={{ cookiecutter.kedro_version }}",
+    "kedro-telemetry~=0.2.0",
+]
+dynamic = ["version"]
+
+[project.scripts]
+{{ cookiecutter.repo_name }} = "{{ cookiecutter.python_package }}.__main__:main"
+
+[project.optional-dependencies]
+docs = [
+    "docutils<0.18.0",
+    "sphinx~=3.4.3",
+    "sphinx_rtd_theme==0.5.1",
+    "nbsphinx==0.8.1",
+    "nbstripout~=0.4",
+    "sphinx-autodoc-typehints==1.11.1",
+    "sphinx_copybutton==0.3.1",
+    "ipykernel>=5.3, <7.0",
+    "Jinja2<3.1.0",
+    "myst-parser~=0.17.2",
+]
+dev = [
+    "black~=22.0",
+    "flake8>=3.7.9, <5.0",
+    "ipython>=7.31.1, <8.0; python_version < '3.8'",
+    "ipython~=8.10; python_version >= '3.8'",
+    "isort~=5.0",
+    "jupyter~=1.0",
+    "jupyterlab_server>=2.11.1, <2.16.0",
+    "jupyterlab~=3.0, <3.6.0",
```

Comment:
Wait are the `ipython` and `jupyter` requirements just dev requirements? They're needed for `kedro ipython` and `kedro jupyter` right? cc: @antonymilne

Reply:
Good point. Not sure about `jupyter` (in fact https://pypi.org/project/jupyter/ is mostly a metapackage, we should be explicitly installing `notebook`, `ipykernel`, or whatever we need).

About `ipython`, most likely:

https://github.com/kedro-org/kedro/blob/41f03d903cda4d4b4f61b01091f494b7b0817e79/kedro/ipython/__init__.py#L12-L14

But do we want a hard dependency on it? It's useful for interactive use, but not so much for deployment I'd say.

Reply:
It's not necessarily useful for deployment, but I do think we have a pretty significant number of people who use both `kedro ipython` and `kedro jupyter notebook`. To me it seems a downgrade on experience if now they have to do an extra installation step to make this commands work. I think it's a good idea to discuss this again later, but I don't think we should hastily make a change like that in this PR.

Reply:
I discussed this with Ivan a while ago and our conclusion was that jupyter requirements should indeed be dev requirements. They are indeed needed for `kedro ipython/jupyter`, but that is part of the development workflow. What defines a core project requirement is that it's need for `kedro run` and would be a requirement when you do `kedro package.`

Note that these commands have a check built in to see if the dependencies are installed:
https://github.com/kedro-org/kedro/blob/41f03d903cda4d4b4f61b01091f494b7b0817e79/kedro/framework/cli/utils.py#L355-L357

What we should do is change what that this message says, but this is where things get a bit tricky (will post in a separate comment about this).

Reply:
I think this is an important discussion but maybe we should decouple it from this transition. We can do this in two parts:

1. Migrate everything to `pyproject.toml`, and keep the `ipython` dependency intact to avoid more churn (this PR)
2. Evaluate whether moving `ipython` to dev dependency requires a deprecation warning of some sort (post 0.19)

What do you folks think?

Reply:
Sorry, essentially what @merelcht said:

> I think it's a good idea to discuss this again later, but I don't think we should hastily make a change like that in this PR.